### PR TITLE
whitelist query parameters

### PR DESCRIPTION
**Pull request by dekkagaijin**

### CODE REVIEW

**Comment by dekkagaijin** [comment]
PTAL @mikedanese

File: metadata/metadata_test.go [hunk 1]
```diff
@@ -18,6 +18,10 @@ var (
 	parseErr         = errors.New("Metadata proxy could not safely parse request.")
 )
 
+func unknownQueryParameterErr(key string) error {
+	return fmt.Errorf("Unrecognized query parameter key: \"%#v\".", key)
```
Comment (mikedanese) [inline metadata/metadata_test.go]: just use %q, also errors shouldn't have periods.

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: metadata/metadata.go [hunk 2]
```diff
@@ -85,8 +104,14 @@ func Filter(req *http.Request) (string, error) {
 		cleanedPath += "/"
 	}
 
+	for key := range req.URL.Query() {
+		if !knownQueryParameterKey[key] {
+			return "", fmt.Errorf("Unrecognized query parameter key: \"%#v\".", key)
```
Comment (mikedanese) [inline metadata/metadata.go]: Same as other comment.

File: metadata/metadata_test.go [hunk 3]
```diff
@@ -58,10 +69,13 @@ func TestFilterURL(t *testing.T) {
...
 		{"/computeMetadata/v1/instance/service-accounts/default/identity", concealedErr, ""},
+		{"/computeMetadata/v1/instance/service-accounts/default/identity?audience=www.example.com&format=full", concealedErr, ""},
 		// Recursive (non-whitelisted).
 		{"/computeMetadata/v1/instance/?recursive=true", recursiveErr, ""},
-		{"/computeMetadata/v1/instance/?something=other&recursive=true", recursiveErr, ""},
-		{"/computeMetadata/v1/instance/?recursive=true&something=other", recursiveErr, ""},
+		{"/computeMetadata/v1/instance/?%72%65%63%75%72%73%69%76%65=true", recursiveErr, ""}, // url-hex-encoded
+		{"/computeMetadata/v1/instance/?alt=text&recursive=true", recursiveErr, ""},
+		{"/computeMetadata/v1/instance/?recursive=true&alt=text", recursiveErr, ""},
+		{"/computeMetadata/v1/instance/?alt=text;recursive=true", recursiveErr, ""},
```
Comment (mikedanese) [inline metadata/metadata_test.go:79]: Can we add a test for `/computeMetadata/v1/instance/?recursive`

File: metadata/metadata_test.go (hunk 1 above)
Comment (dekkagaijin) [reply metadata/metadata_test.go]: I specifically used %#v so that non-unicode characters show up as byte literals

Comment (dekkagaijin) [reply metadata/metadata_test.go]: looks like `%#q` does the right thing

File: metadata/metadata_test.go (hunk 3 above)
Comment (dekkagaijin) [reply metadata/metadata_test.go:79]: done

File: metadata/metadata.go (hunk 2 above)
Comment (dekkagaijin) [reply metadata/metadata.go]: done

File: metadata/metadata_test.go (hunk 1 above)
Comment (dekkagaijin) [reply metadata/metadata_test.go]: done!

File: metadata/metadata.go [hunk 4]
```diff
@@ -85,8 +104,14 @@ func Filter(req *http.Request) (string, error) {
 		cleanedPath += "/"
 	}
 
+	for key := range req.URL.Query() {
+		if !knownQueryParameterKey[key] {
+			return "", fmt.Errorf("Unrecognized query parameter key: %#q.", key)
```
Comment (mikedanese) [inline metadata/metadata.go]: Still has a period. Is that intentional?

**Comment by mikedanese** [comment]
LGTM

File: OWNERS [hunk 5]
```diff
@@ -2,7 +2,9 @@ approvers:
 - ihmccreery
 - cjcullen
 - destijl
+- dekkagaijin
 reviewers:
 - ihmccreery
```
Comment (mikedanese) [inline OWNERS]: Can you add me and remove Ike?

File: metadata/metadata.go (hunk 4 above)
Comment (dekkagaijin) [reply metadata/metadata.go]: whoops, sorry, done

File: OWNERS (hunk 5 above)
Comment (dekkagaijin) [reply OWNERS]: done

**Pull request merged**